packages/realm-react/src/useAuthOperation.tsx

```diff
@@ -0,0 +1,72 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { useCallback } from "react";
+import { AuthError, AuthOperationName, OperationState } from "./types";
+import { useAuthResult } from "./AppProvider";
+
+export function useAuthOperation<Args extends unknown[], Result>({
+  operation,
+  operationName,
+  onSuccess = () => undefined,
+}: {
+  operation: (...args: Args) => Promise<Result | void>;
+  operationName: AuthOperationName;
+  onSuccess?: (...args: Args) => void;
+}) {
+  const [result, setResult] = useAuthResult();
+
+  return useCallback<(...args: Args) => ReturnType<typeof operation>>(
+    (...args) => {
+      if (result.pending) {
+        return Promise.reject("Another authentication operation is already in progress");
+      }
+
+      setResult({
+        state: OperationState.Pending,
+        pending: true,
+        success: false,
+        error: undefined,
+        currentOperation: operationName,
+      });
+      return operation(...args)
+        .then((res) => {
+          setResult({
+            state: OperationState.Success,
+            pending: false,
+            success: true,
+            error: undefined,
+            currentOperation: operationName,
+          });
+          onSuccess(...args);
+          return res;
+        })
+        .catch((error) => {
```

Comment:
I came to the conclusion that automatic login after register was too side-effecty.  I believe it would only be nice for the templates, but not for a "real-world" authentication experience.  With this being removed, the `onSuccess` is no longer needed.